lib/matplotlib/ticker.py

```diff
@@ -2442,35 +2449,70 @@ def tick_values(self, vmin, vmax):
         else:
             subs = self._subs
 
-        # Get decades between major ticks.
-        stride = (max(math.ceil(numdec / (numticks - 1)), 1)
-                  if mpl.rcParams['_internal.classic_mode'] else
-                  numdec // numticks + 1)
-
-        # if we have decided that the stride is as big or bigger than
-        # the range, clip the stride back to the available range - 1
-        # with a floor of 1.  This prevents getting axis with only 1 tick
-        # visible.
-        if stride >= numdec:
-            stride = max(1, numdec - 1)
-
-        # Does subs include anything other than 1?  Essentially a hack to know
-        # whether we're a major or a minor locator.
-        have_subs = len(subs) > 1 or (len(subs) == 1 and subs[0] != 1.0)
-
-        decades = np.arange(math.floor(log_vmin) - stride,
-                            math.ceil(log_vmax) + 2 * stride, stride)
-
-        if have_subs:
-            if stride == 1:
-                ticklocs = np.concatenate(
-                    [subs * decade_start for decade_start in b ** decades])
+        # Get decades between major ticks.  Always return an extra tick outside
+        # the lower and the upper limit: QuadContourSet._autolev relies on
+        # this.
+        if mpl.rcParams["_internal.classic_mode"]:  # keep historic formulas
+            stride = max(math.ceil((nd - 1) / (nt - 1)), 1)
+            decades = np.arange(emin - stride, emax + stride + 1, stride)
+        else:
+            # Find the largest number of ticks, no more than the requested
+            # number, that can actually be drawn (e.g., with 9 decades ticks,
+            # no stride yields 7 ticks).  For a given value of the stride *s*,
+            # there are either floor(nd/s) or ceil(nd/s) ticks depending on the
+            # offset.  Pick the smallest stride such that floor(nd/s) < nt,
+            # i.e. nd/s < nt+1, then re-set nt to ceil(...) if acceptable, else
+            # to floor(...) (which must then equal the original nt, i.e. nt is
+            # kept unchanged).
+            stride = nd // (nt + 1) + 1
```

Comment:
I'm not following this formula. Isn't it that, assuming we accept a fractional stride $nd - 1 = s * (nt -1)$, i.e. $s = (nd - 1) / (nt - 1)$ and then we floor that to get the next integer stride that does not make $s * (nt -1)$ larger than before.